lms/djangoapps/courseware/access.py

```diff
@@ -507,14 +538,16 @@ def can_load():
         students to see modules.  If not, views should check the course, so we
         don't have to hit the enrollments table on every module load.
         """
-        return (
-            not descriptor.visible_to_staff_only
+        response = (
+            _visible_to_nonstaff_users(descriptor)
             and _has_group_access(descriptor, user, course_key)
-            and (
-                'detached' in descriptor._class_tags  # pylint: disable=protected-access
+            and
+            (
+                _has_detached_class_tag(descriptor)
                 or _can_access_descriptor_with_start_date(user, descriptor, course_key)
             )
-        ) or _has_staff_access_to_descriptor(user, descriptor, course_key)
+        )
+        return response if response else _has_staff_access_to_descriptor(user, descriptor, course_key)
```

Comment:
No worries.

What this shows, however, is that this design is not completely ideal. It now makes it more difficult to write the access control logic, since we want particular error objects returned. That is now outside the scope of this PR.

However, since we are providing a common access-control framework, this should be addressed. I will wait and see whether the access.py changes resulting from @mekkz's Transformers work will make this easier.